src/python/pants/engine/legacy/graph.py

```diff
@@ -719,14 +718,23 @@ def _eager_fileset_with_spec(
 async def provenanced_addresses_from_filesystem_specs(
   filesystem_specs: FilesystemSpecs,
 ) -> ProvenancedBuildFileAddresses:
-  snapshot = await Get[Snapshot](PathGlobs, filesystem_specs.to_path_globs())
-  owners = await Get[Owners](OwnersRequest(sources=snapshot.files))
-  # TODO(#9055): do we care about preserving the original provenance? This would be tricky to
-  # implement. For now, we use a no-op `FilesystemSpec("")`.
-  return ProvenancedBuildFileAddresses(
-    ProvenancedBuildFileAddress(build_file_address=bfa, provenance=FilesystemSpec(""))
-    for bfa in owners.addresses
+  """Find the owner(s) for each FilesystemSpec."""
+  pathglobs_per_include = (
+    filesystem_specs.path_globs_for_spec(spec) for spec in filesystem_specs.includes
+  )
+  snapshot_per_include = await MultiGet(
+    Get[Snapshot](PathGlobs, pg) for pg in pathglobs_per_include
   )
+  owners_per_include = await MultiGet(
+    Get[Owners](OwnersRequest(sources=snapshot.files)) for snapshot in snapshot_per_include
+  )
+  result: List[ProvenancedBuildFileAddress] = []
+  for spec, owners in zip(filesystem_specs.includes, owners_per_include):
+    result.extend(
+      ProvenancedBuildFileAddress(build_file_address=bfa, provenance=spec)
+      for bfa in owners.addresses
+    )
```

Comment:
If the goal of this is to "preserve the original provenance" to fix #9055, would it be useful to leave a comment here describing that?

Reply:
I'll update the docstring to mention "preserving the original provenance". I had that originally but didn't like the grammar so got rid of it 🙈